design/name-system.md

```diff
@@ -0,0 +1,103 @@
+* **[Multiaddr]s** is an extensible format describing a network address through multiple layers and protocols.
+  * Example representation for a machine at IP address 10.20.30.1 with TCP port 20000: `/ip4/10.20.30.1/tcp/20000`
+* **[PeerId]s** are [multihash]ed public keys. In NNS, PeerIds are derived from keys unrelated to spheres, and used to identify and communicate with DHT nodes in the network.
+* **Content Identifiers** ([CID]s) are [multihash]ed [content addresses](https://docs.ipfs.tech/concepts/content-addressing/), where content can be identified by an immutable address. CIDs used in NNS are all CIDv1.
+
+## Overview
+
+A sphere in Noosphere contains internal data for an address book, where other sphere's identities ([DID]s) can be stored. DIDs are hardly memorizable, so they're stored with a name, like "Alice". When Bob wants to reference Alice's notes on dogs from his sphere in subtext, it'd look like `"@Alice/dogs"`. The "Alice" entry in the address book maps to a DID, like `did:key:z6MkoE19WHXJzpLqkxbGP7uXdJX38sWZNUWwyjcuCmjhPpUP`, but a [CID] address is needed in order to fetch, ideally the latest, contents from Alice's sphere. This is where the Noosphere Name System comes in.
+
+Bob's gateway connects to the NNS, asking peers if anyone has an address for Alice's key. The address can be found from the network, and using [UCAN], can be validated that it was Alice publishing it from Alice's gateway, or maybe even from another peer on the network that has no idea who Alice is. Once the address is retrieved, Alice's sphere's content can be fetched and remixed.
+
+Later, Bob wants Alice's latest content. Because these addresses are content-addressed CIDs, they're immutable, meaning everytime a sphere's content changes, it has a new address. Bob's gateway connects to the NNS again, looking for a newer address for Alice's sphere, comparing the UCAN timestamps from previous and current results for Alice's sphere.
+
+Additionally, Bob's gateway is propagating Bob's sphere's latest revision address when the sphere changes for others to find.
+
+## Network Spec
+
+The NNS network adheres to [libp2p]'s [Kademlia DHT Spec](https://github.com/libp2p/specs/blob/master/kad-dht/README.md)\*, and defines a set of expected values, validations, and recommended configurations.
+
+Nodes in the libp2p DHT are identified by [PeerId]. These **SHOULD** be generated from [Ed25519] keys.
+
+\* *Technically, libp2p Kademlia DHT specifies that keys in the network should be CIDs. NNS uses DIDs as keys.*
+
+### Protocols
+
+The DHT nodes **MUST** implement the libp2p [Identify] spec.
+
+### DHT Methods 
+
+All methods in the underlying spec **MUST** be implemented, with cavaets below.
+
+The network uses Value Records (using `GET_VALUE`, `PUT_VALUE` DHT methods) where the key **SHOULD** be a sphere's public [Ed25519] key encoded as a [DID] UTF-8 string, and the corresponding value **SHOULD** be a [UCAN] token encoded as a [JWT] string.
```

Comment:
The value could be a CID itself in the future, resolving into a UCAN. Should be straight forward to implement this. I think defining both forms is reasonable

Reply:
Although we discussed this possibility, worth noting that there is a likely performance advantage to making it a JWT and not a CID pointing to a JWT: the top-level record is likely to change a lot, but its associated proofs are likely to stay the same for long time. So, by making the record a JWT, we always send around the part that changes every time, and can rely on caching to take care of the parts that don't change much.

Reply:
great points, will note that!